src/Persistence/MassTransit.MongoDbIntegration/MongoDbIntegration/Outbox/MongoDbOutboxContextFactory.cs

```diff
@@ -51,38 +48,38 @@ async Task<bool> Execute()
                             Received = DateTime.UtcNow,
                             ReceiveCount = 0
                         };
+                        isInsert = true;
+                    }
 
-                        await inboxStateCollection.InsertOne(inboxState, context.CancellationToken).ConfigureAwait(false);
+                    inboxState.LockToken = ObjectId.GenerateNewId();
 
-                        continueProcessing = true;
-                    }
-                    else
+                    if (updateReceiveCount)
                     {
-                        if (updateReceiveCount)
-                        {
-                            inboxState.ReceiveCount++;
-                            await inboxStateCollection.FindOneAndReplace(filter, inboxState, context.CancellationToken).ConfigureAwait(false);
-                        }
-
+                        inboxState.ReceiveCount++;
                         updateReceiveCount = false;
+                    }
 
-                        var outboxContext = new MongoDbOutboxConsumeContext<T>(context, options, inboxState, _dbContext);
+                    var outboxContext = new MongoDbOutboxConsumeContext<T>(context, options, inboxState, _dbContext);
 
-                        await next.Send(outboxContext).ConfigureAwait(false);
+                    await next.Send(outboxContext).ConfigureAwait(false);
 
-                        inboxState.Version++;
+                    inboxState.Version++;
 
-                        FilterDefinition<InboxState> updateFilter = builder.Eq(x => x.MessageId, messageId) & builder.Eq(x => x.ConsumerId, options.ConsumerId)
+                    if (isInsert)
+                        await inboxStateCollection.InsertOne(inboxState, context.CancellationToken).ConfigureAwait(false);
+                    else
+                    {
+                        FilterDefinition<InboxState> updateFilter = builder.Eq(x => x.MessageId, messageId) &
+                            builder.Eq(x => x.ConsumerId, options.ConsumerId)
                             & builder.Lt(x => x.Version, inboxState.Version);
 
                         await inboxStateCollection.FindOneAndReplace(updateFilter, inboxState, context.CancellationToken).ConfigureAwait(false);
```

Comment:
We should use ReplaceOne here, because there is no noeed to retrieve the updated document.